samples/cpp/tutorial_code/videoio/astra/Astra.cpp

```diff
@@ -0,0 +1,158 @@
+    }
+
+    // Set color stream size equal to depth stream's ones
+    //colorStream.set(CAP_PROP_FRAME_WIDTH, depthStream.get(CAP_PROP_FRAME_WIDTH));
+    //colorStream.set(CAP_PROP_FRAME_HEIGHT, depthStream.get(CAP_PROP_FRAME_HEIGHT));
+    //! [Setup streams]
+    // Set color and depth stream parameters
+    colorStream.set(CAP_PROP_FRAME_WIDTH,  640);
+    colorStream.set(CAP_PROP_FRAME_HEIGHT, 480);
+    depthStream.set(CAP_PROP_FRAME_WIDTH,  640);
+    depthStream.set(CAP_PROP_FRAME_HEIGHT, 480);
+    depthStream.set(CAP_PROP_OPENNI2_MIRROR, 0);
+    //! [Setup streams]
+
+    // Print color stream parameters
+    cout << "Color stream: "
+         << colorStream.get(CAP_PROP_FRAME_WIDTH) << "x" << colorStream.get(CAP_PROP_FRAME_HEIGHT)
+         << " @" << colorStream.get(CAP_PROP_FPS) << " fps" << endl;
+
+    //! [Get properties]
+    // Print depth stream parameters
+    cout << "Depth stream: "
+         << depthStream.get(CAP_PROP_FRAME_WIDTH) << "x" << depthStream.get(CAP_PROP_FRAME_HEIGHT)
+         << " @" << depthStream.get(CAP_PROP_FPS) << " fps" << endl;
+    //! [Get properties]
+
+    //! [Read streams]
+    // Start two reading threads
+    std::mutex mtx;
+    std::condition_variable dataReady;
+    bool isFinish = false;
+
+    std::thread depthReader([&depthStream, &dataReady, &isFinish]{
+        while (!isFinish)
+        {
+            // Grab new frame
+            bool frameReady = depthStream.grab();
+            if (frameReady)
+                dataReady.notify_all();
+        }
+    });
+
+    std::thread colorReader([&colorStream, &dataReady, &isFinish]{
+        while (!isFinish)
+        {
+            // Grab new frame
+            bool frameReady = colorStream.grab();
+            if (frameReady)
+                dataReady.notify_all();
+        }
+    });
+    //! [Read streams]
+
+    int errCode = 0;
+    Mat colorFrame, irFrame, depthMap, disparityMap;
+    while (true)
+    {
+        std::unique_lock<std::mutex> lk(mtx);
+        dataReady.wait(lk);
+
+        //! [Decode depth]
+        // Decode grabbed depth data
+        depthStream.retrieve(depthMap, CAP_OPENNI_DEPTH_MAP);
+        if (depthMap.empty())
+        {
+            cerr << "ERROR: Failed to grab depth map from depth stream" << endl;
+            errCode = 2;
+            break;
+        }
+        //! [Decode depth]
+
+        depthStream.retrieve(disparityMap, CAP_OPENNI_DISPARITY_MAP);
+        if (disparityMap.empty())
+        {
+            cerr << "ERROR: Failed to grab disparity map from depth stream" << endl;
+            errCode = 2;
+            break;
+        }
+
+        depthStream.retrieve(irFrame, CAP_OPENNI_IR_IMAGE);
+        if (irFrame.empty())
+        {
+            cerr << "ERROR: Failed to grab IR frame from depth stream" << endl;
+            errCode = 2;
+            break;
+        }
+
+        //! [Decode color]
+        // Decode grabbed color data
+        colorStream.retrieve(colorFrame);
```

Comment:
Threads should not call `.grab()` during `.retrieve()` calls from another thread.
- OpenCV API is not thread safe in this way (except storage containers partially, like Mat/UMat). Threads can't invoke methods of the same object instance without extra external synchronization.
- Capturing garbage may appear if buffers queue is full.

Reply:
Should be fixed in the latest update.